subsys/bluetooth/host/gatt.c

```diff
@@ -4086,6 +4107,34 @@ int bt_gatt_subscribe(struct bt_conn *conn,
 	return 0;
 }
 
+int bt_gatt_add_subscription(uint8_t id, const bt_addr_le_t *addr,
+			     struct bt_gatt_subscribe_params *params)
+{
+	struct gatt_sub *sub;
+	struct bt_gatt_subscribe_params *tmp;
+
+	__ASSERT(params && params->notify,  "invalid parameters\n");
+	__ASSERT(params->value, "invalid parameters\n");
+	__ASSERT(params->ccc_handle, "invalid parameters\n");
+
+	sub = gatt_sub_add_by_addr(id, addr);
+	if (!sub) {
+		return -ENOMEM;
+	}
+
+	/* Lookup existing subscriptions */
+	SYS_SLIST_FOR_EACH_CONTAINER(&sub->list, tmp, node) {
```

Comment:
Does this need to be insertion/removal safe?

Reply:
I don't think so. The call to gatt_sub_remove will AFAICT not do anything. Because params is NULL in the call, and the sub->list cannot be empty (since we found a duplicate)

If you mean that `bt_gatt_unsubcsribe` could be call from another thread then we should fix `bt_gatt_subscribe` as well.
@Vudentz Should we switch to safe foreach here?

Reply:
Yep, I guess both subscribe and unsubscribe should be using the safe version, that said that is only safe for the foreach not sure if @carlescufi was talking about have a lock or something to prevent other threads to change the list.